### PR TITLE
PR #43: Shooter Without Pivot (ShooterCommands to develop)

### PR DESCRIPTION
**Pull request by Irishumanoid**
It shoots notes without pivot. Indexing motor, 2 flywheel motors, and shield, run by commands.
Trello: https://trello.com/c/Vs9Oh51x/94-pr-42-shooter-without-pivot-shootercommands-to-develop

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -46,7 +46,7 @@ def deployArtifact = deploy.targets.roborio.artifacts.frcJava
 wpi.java.debugJni = false
 
 // Set this to true to enable desktop support.
-def includeDesktopSupport = false
+def includeDesktopSupport = true
```
Comment (garrettwitowski) [inline build.gradle]: Why this change?

File: src/main/deploy/pathplanner/autos/AmpFromBottom.auto [hunk 2]
```diff
@@ -0,0 +1,56 @@
+{
```
Comment (garrettwitowski) [inline src/main/deploy/pathplanner/autos/AmpFromBottom.auto]: It seems like this needs to be rebased on top of develop?

File: src/main/java/frc/robot/RobotContainer.java [hunk 3]
```diff
@@ -4,61 +4,62 @@
...
 
-  public boolean triggerPressed() {
-    if (m_driverController.getLeftTriggerAxis() != 0
-        || m_driverController.getRightTriggerAxis() != 0) {
-      return true;
-    }
-    return false;
+    // triggers for extending and retracting shield manually
+    new Trigger(() -> m_operatorController.getRawButton(Bindings.kExtendShield))
+        .onTrue(new ActuateShield(m_shooter, false));
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/RobotContainer.java:108]: nit: add a comment next to the boolean here to indicate what value that's setting (it's not clear if you don't have an IDE that provides that info for you). This comment is applicable in several places.

File: src/main/java/frc/robot/RobotContainer.java [hunk 4]
```diff
@@ -4,61 +4,62 @@
...
+    // angle on 8-directional button
+    m_autoAim = new POVButton(m_operatorController, 0);
+    m_speakerAim = new POVButton(m_operatorController, 90);
+
+    // just shoot on trigger
+    new Trigger(() -> m_operatorController.getRawButton(Bindings.kShoot))
+        .whileTrue(new Shoot(m_intake, false));
+    new Trigger(() -> m_operatorController.getRawButton(Bindings.kShootReverse))
+        .whileTrue(new Shoot(m_intake, true));
+    // aim amp (velocity only rn)
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/RobotContainer.java]: nit: no need to specify that it's velocity only currently -- you're leaking implementation details to a file that otherwise doesn't care about them.

File: src/main/java/frc/robot/commands/VisionTranslateCommand.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import edu.wpi.first.wpilibj2.command.Command;
+import frc.robot.constants.RobotConfig.DriveConfig.TranslateConfig;
+import frc.robot.constants.RobotConstants.DriveConstants;
+import frc.robot.subsystems.drive.Drivetrain;
+import frc.robot.subsystems.vision.Vision;
+import frc.utils.Vector;
+
+public class VisionTranslateCommand extends Command {
+
+  // private VisionSubsystem vision;
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/commands/VisionTranslateCommand.java]: Delete dead code

File: src/main/java/frc/robot/commands/VisionTranslateCommand.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    addRequirements(vision, drive);
+
+    forwardController.setIntegratorRange(TranslateConfig.minIntegral, TranslateConfig.maxIntegral);
+  }
+
+  @Override
+  public void execute() {
+    double forwardSpeed = 0.0;
+
+    if (vision.getHasTarget()) {
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/commands/VisionTranslateCommand.java:42]: Has all this vision code been tested?

File: src/main/java/frc/robot/commands/shooter/ActuateShield.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  }
+
+  @Override
+  public void initialize() {
+    initTime = Timer.getFPGATimestamp();
+  }
+
+  @Override
+  public void execute() {
+    System.out.println("executing shield");
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/commands/shooter/ActuateShield.java]: Remove debug

File: src/main/java/frc/robot/commands/shooter/Aim.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    m_vision = eyes;
+    m_type = null;
+    addRequirements(m_shooter, m_vision);
+  }
+
+  @Override
+  public void initialize() {
+    initTime = Timer.getFPGATimestamp();
+    if (m_type == null) {
+      if (m_vision.getHasTarget()) {
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/commands/shooter/Aim.java:39]: Why the vision stuff? Has it been tested?

File: src/main/java/frc/robot/commands/shooter/Aim.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+                m_vision.getDistToTarget() * Math.atan(desiredAngle),
+                Units.Radians.of(desiredAngle));
+
+        m_shooter.runFlywheel(m_shooter.convertToRPM(desiredVelocity.magnitude()));
+      }
+    } else {
+      switch (m_type) {
+        case AMP:
+          desiredAngle = ShooterConfig.kAmpAngle;
+          // desiredVelocity = getVelocity(ShooterConfig.AmpHeight);
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/commands/shooter/Aim.java]: Delete dead code

File: src/main/java/frc/robot/commands/shooter/Aim.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+          desiredVelocity = ShooterConfig.ampVelocity;
+          break;
+        case SPEAKER:
+          desiredAngle = ShooterConfig.kSpeakerAngle;
+          desiredVelocity = getVelocity(ShooterConfig.SpeakerHeight);
+          break;
+        case TRAP:
+          desiredAngle = ShooterConfig.kTrapAngle;
+          // desiredVelocity = getVelocity(ShooterConfig.TrapHeight);
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/commands/shooter/Aim.java]: Dead code

File: src/main/java/frc/robot/commands/shooter/Aim.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+      }
+      m_shooter.runFlywheel(desiredVelocity);
+    }
+  }
+
+  @Override
+  public void execute() {
+    double timeDiff = Timer.getFPGATimestamp() - initTime;
+    if (m_type == FieldElement.AMP
+        || m_type == FieldElement.TRAP && timeDiff < ShooterConfig.kShieldTime) {
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/commands/shooter/Aim.java]: I see this timeDiff stuff a lot -- why is it here?

File: src/main/java/frc/robot/commands/shooter/Shoot.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+  }
+
+  @Override
+  public void end(boolean interrupted) {
+    m_intake.stopFeedNote();
+  }
+
+  @Override
+  public boolean isFinished() {
+    return Timer.getFPGATimestamp() - timer > RobotConfig.ShooterConfig.kReleaseTime;
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/commands/shooter/Shoot.java]: Why not just "shoot while trigger pressed" and eliminate the need for timing?

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java [hunk 13]
```diff
@@ -1,21 +1,22 @@
...
-import edu.wpi.first.math.kinematics.SwerveDriveOdometry;
 import edu.wpi.first.math.kinematics.SwerveModulePosition;
 import edu.wpi.first.units.*;
 import edu.wpi.first.wpilibj.PowerDistribution;
 import edu.wpi.first.wpilibj.Timer;
 import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
 import edu.wpi.first.wpilibj2.command.SubsystemBase;
-import frc.robot.constants.RobotConfig.DriveConfig;
-import frc.robot.constants.RobotConstants.DriveConstants;
+import frc.robot.constants.RobotConfig.*;
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: Should not import .* as a rule of thumb -- please change to specific imports.

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java [hunk 14]
```diff
@@ -89,33 +106,33 @@ public Drivetrain() {
     m_timer.start();
     m_prevTime = m_timer.get();
 
-    m_odometry =
-        new SwerveDriveOdometry(
+    m_swervePoseEstimator =
+        new SwerveDrivePoseEstimator(
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: Why are we changing this as part of shooter?

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java [hunk 15]
```diff
@@ -89,33 +106,33 @@ public Drivetrain() {
...
-            });
+            SwerveModulePositions,
+            new Pose2d());
 
     m_powerDistribution.clearStickyFaults();
     SmartDashboard.putNumber("driveVelocity", 0);
   }
 
-  /** runs the periodic functionality of the drivetrain */
+  /** runs the periodic functionality of the drivetrain * */
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: nit: extra space

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java [hunk 16]
```diff
@@ -89,33 +106,33 @@ public Drivetrain() {
...
-          m_rearRight.getPosition(),
-        });
+    // Update the pose estimator in the periodic block
+    m_swervePoseEstimator.update(
+        Rotation2d.fromRadians(-getGyroAngle().in(Units.Radians)), SwerveModulePositions);
+
+    // Update the pose estimator with data from the vision pose estimator
+    Pose2d visPose = m_vision.getEstimatedPose2d();
+    if (visPose != null) {
+      // var pipelineResult = m_vision.getCam().getLatestResult();
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: Delete dead code

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java [hunk 17]
```diff
@@ -89,33 +106,33 @@ public Drivetrain() {
...
+      // var resultTimestamp = pipelineResult.getTimestampSeconds();
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: Dead code

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java [hunk 18]
```diff
@@ -89,33 +106,33 @@ public Drivetrain() {
...
+
+      /*m_swervePoseEstimator.addVisionMeasurement(
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: Dead code

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java [hunk 19]
```diff
@@ -89,33 +106,33 @@ public Drivetrain() {
...
+      m_vision.getEstimatedPose2d(), Timer.getFPGATimestamp());*/
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: Ditto

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 20]
```diff
@@ -1,18 +1,210 @@
...
+  }
+
+  public void stopFlywheel() {
+    m_topFlywheelMotor.stopMotor();
+    ;
+  }
+
+  public void zeroEncoders() {
+    m_topFlywheelEncoder.setPosition(0);
+    m_bottomFlywheelEncoder.setPosition(0);
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:167]: nit: Does the position really matter for these flywheels?

**Review by garrettwitowski — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/VisionTranslateCommand.java (hunk 5 above)
Comment (continuumsrc) [reply src/main/java/frc/robot/commands/VisionTranslateCommand.java]: Deleted.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 20 above)
Comment (continuumsrc) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java:167]: No, but it's good practice.

**Review by mojo626 — APPROVED** [review]

File: src/main/java/frc/robot/commands/shooter/Aim.java (hunk 8 above)
Comment (Irishumanoid) [reply src/main/java/frc/robot/commands/shooter/Aim.java:39]: The constructor for vision align isn't used currently, so it doesn't break anything. Currently unused but want to push so we can test at comp easier.

File: build.gradle (hunk 1 above)
Comment (garrettwitowski) [reply build.gradle]: Again, why this change?

File: src/main/java/frc/robot/RobotContainer.java [hunk 21]
```diff
@@ -14,43 +10,63 @@
 import edu.wpi.first.wpilibj2.command.Command;
 import edu.wpi.first.wpilibj2.command.Commands;
 import edu.wpi.first.wpilibj2.command.RunCommand;
+import edu.wpi.first.wpilibj2.command.button.POVButton;
 import edu.wpi.first.wpilibj2.command.button.Trigger;
 import frc.robot.commands.BasicDriveCommand;
 import frc.robot.commands.intake.RunIntake;
-import frc.robot.commands.intake.RunIntakeReversed;
+import frc.robot.commands.shooter.*;
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/RobotContainer.java]: Please do not import .*

File: src/main/java/frc/robot/RobotContainer.java [hunk 22]
```diff
@@ -14,43 +10,63 @@
...
 import frc.robot.constants.RobotConfig;
+import frc.robot.constants.RobotConfig.*;
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/RobotContainer.java]: Ditto

File: src/main/java/frc/robot/subsystems/drive/Drivetrain.java [hunk 23]
```diff
@@ -343,7 +366,10 @@ private void limitDirectionSlewRate(Vector spdVec) {
 
     // Calculate the direction slew rate based on an estimate of the lateral acceleration
     double directionSlewRate;
-    if (m_currentTranslationMag > 1e-4) {
+    // if very close to zero but not exactly zero, there is no in division by zero due to floating
+    // point precision errors
+    if (m_currentTranslationMag != 0) {
```
Comment (garrettwitowski) [inline src/main/java/frc/robot/subsystems/drive/Drivetrain.java]: Don't we want the 1e-4 check?

**Review by garrettwitowski — DISMISSED** [review]

File: build.gradle (hunk 1 above)
Comment (TurtleMeds) [reply build.gradle]: https://docs.wpilib.org/en/stable/docs/software/wpilib-tools/robot-simulation/introduction.html

Good question. The link above says it may have unintended consequences so we should probably disable it.

it looks like it was enabled when I added the camera name and initialized vision. no idea why that happened

File: src/main/java/frc/robot/commands/shooter/Aim.java (hunk 8 above)
Comment (garrettwitowski) [reply src/main/java/frc/robot/commands/shooter/Aim.java:39]: We really shouldn't push to main/develop if it's untested

File: src/main/java/frc/robot/commands/VisionTranslateCommand.java [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+  private PIDController forwardController;
+
+  public VisionTranslateCommand(Vision vision, Drivetrain drive, XboxController controller) {
+    this.vision = vision;
+    this.drive = drive;
+
+    this.controller = controller;
+    SmartDashboard.putNumber(TranslateConfig.kPKey, TranslateConfig.kP);
+    SmartDashboard.putNumber(TranslateConfig.kIKey, TranslateConfig.kI);
+    SmartDashboard.putNumber(TranslateConfig.kDKey, TranslateConfig.kD);
```
Comment (ChainedGCH) [inline src/main/java/frc/robot/commands/VisionTranslateCommand.java:29]: Do we still have test mode/having shuffleboard set PID values?

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 25]
```diff
@@ -1,18 +1,210 @@
...
-  public void simulationPeriodic() {
-    // This method will be called once per scheduler run during simulation
+  void testPeriodic() {
+    SmartDashboard.putNumber("Shooter/top flywheel output", m_topFlywheelMotor.getAppliedOutput());
+    SmartDashboard.putNumber(
+        "Shooter/bottom flywheel output", m_bottomFlywheelMotor.getAppliedOutput());
+    double flywheelRPM =
+        SmartDashboard.getNumber("Shooter/Flywheel RPM", m_topFlywheelEncoder.getVelocity());
+
+    double pval = SmartDashboard.getNumber("flywheel p", 0.1);
```
Comment (ChainedGCH) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:120]: Why is the string not a key?

**Review by ChainedGCH — CHANGES_REQUESTED** [review]
Please add a note about Vision not doing anything in the actual code

**Review by ChainedGCH — APPROVED** [review]
Just in the future, please add functionality to vision or Please add a note about Vision not doing anything in the actual code

**Review by continuumsrc — APPROVED** [review]

File: src/main/java/frc/robot/commands/shooter/Aim.java (hunk 11 above)
Comment (Irishumanoid) [reply src/main/java/frc/robot/commands/shooter/Aim.java]: removed, not necessary

File: src/main/java/frc/robot/commands/VisionTranslateCommand.java (hunk 24 above)
Comment (Irishumanoid) [reply src/main/java/frc/robot/commands/VisionTranslateCommand.java:29]: yes
